mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,307 @@
+                                       const scf::YieldOp &SCFRegionYieldOp,
+                                       std::shared_ptr<SmallVector<Operation *>> TapeOps,
+                                       SmallVector<Value> &RetValues)
+    {
+        SmallPtrSet<Operation *, 16> exceptions;
+        exceptions.insert(SCFRegionYieldOp);
+        for (auto op : *TapeOps) {
+            exceptions.insert(op);
+        }
+
+        for (size_t i = 0; i < RetValues.size(); i++) {
+            RetValues[i].replaceAllUsesExcept(executeRegionOp->getResults()[i], exceptions);
+        }
+    } // propagateSCFRetValsDownstream()
+
+    LogicalResult createTapeFunction(std::shared_ptr<SmallVector<Operation *>> TapeOps,
+                                     SmallVector<Value> &NecessaryValuesFromEarlierTapes,
+                                     IRRewriter &builder, const unsigned int &tape_number,
+                                     func::FuncOp &OriginalMultitapeFunc,
+                                     SmallVector<FailureOr<func::FuncOp>> &OutlinedFuncs)
+    {
+        assert(TapeOps); // nullptr protection
+
+        // 1. Identify the necessary return values
+        SmallVector<Value> RetValues;
+        SmallVector<mlir::Type> RetTypes;
+        getNecessaryTapeReturns(RetValues, TapeOps, NecessaryValuesFromEarlierTapes);
+        for (Value v : RetValues) {
+            RetTypes.push_back(v.getType());
+        }
+
+        // 2. Create a scf.execute_region and wrap it around the tape ops
+        // The scf.execute_region needs to yield (aka return) the values identified
+        std::pair<scf::ExecuteRegionOp, scf::YieldOp> WrappingRegionAndYield =
+            wrapTapeOpsInSCFRegion(TapeOps, RetValues, RetTypes, builder,
+                                   OriginalMultitapeFunc->getLoc());
+        scf::ExecuteRegionOp executeRegionOp = WrappingRegionAndYield.first;
+        scf::YieldOp SCFRegionYieldOp = WrappingRegionAndYield.second;
+
+        // 3. The scf outliner automatically captures values not defined in the
+        // to-be-outlined function body and transforms them to the outlined function's
+        // arguments, but it does not automatically propagate the returned value
+        // downstream to replace the previously used values. This we need to do manually.
+
+        // At this stage, the scf.yield values ARE the ones to be replaced by
+        // the scf region's result values, since we never replaced anything yet
+        // Of course, don't replace the use in the region itself!
+        propagateSCFRetValsDownstream(executeRegionOp, SCFRegionYieldOp, TapeOps, RetValues);
+
+        // 4. Outline the region
+        func::CallOp call;
+        FailureOr<func::FuncOp> outlined = outlineSingleBlockRegion(
+            builder, OriginalMultitapeFunc->getLoc(), executeRegionOp.getRegion(),
+            OriginalMultitapeFunc.getSymName().str() + "_tape_" + std::to_string(tape_number),
```

Comment:
To get unique names use [`SymbolTable::renameToUnique`](https://mlir.llvm.org/doxygen/classmlir_1_1SymbolTable.html#ad3560c81d1bb5f7376a9c2320a6744d1)

Reply:
Personally I would actually prefer explicit control of the names; otherwise the names generated by this method might not be constant and could make testing difficult

Reply:
I ended up doing the renaming manually by checking if the custom name of the outlined function already exists in the symbol table before the outline actually happens. 

`renameToUnique` was a bit too difficult to use together with the outliner because the outliner updates the module symbol table.